craft_parts/sources/rpm_source.py

```diff
@@ -0,0 +1,115 @@
+#
+# This program is free software; you can redistribute it and/or
+# modify it under the terms of the GNU Lesser General Public
+# License version 3 as published by the Free Software Foundation.
+#
+# This program is distributed in the hope that it will be useful,
+# but WITHOUT ANY WARRANTY; without even the implied warranty of
+# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
+# Lesser General Public License for more details.
+#
+# You should have received a copy of the GNU Lesser General Public License
+# along with this program.  If not, see <http://www.gnu.org/licenses/>.
+
+"""The RPM source handler."""
+
+import logging
+import os
+import subprocess
+import tarfile
+from pathlib import Path
+from typing import List, Optional
+
+from overrides import override
+
+from craft_parts.dirs import ProjectDirs
+
+from . import errors
+from .base import FileSourceHandler
+
+logger = logging.getLogger(__name__)
+
+
+class RpmSource(FileSourceHandler):
+    """The "rpm" file source handler."""
+
+    _invalid_tags = (
+        "source-tag",
+        "source-commit",
+        "source-branch",
+        "source-depth",
+        "source-submodules",
+    )
+    _source_type = "rpm"
+
+    def __init__(
+        self,
+        source: str,
+        part_src_dir: Path,
+        *,
+        cache_dir: Path,
+        source_tag: None = None,
+        source_commit: None = None,
+        source_branch: None = None,
+        source_checksum: Optional[str] = None,
+        source_submodules: None = None,
+        source_depth: None = None,
+        project_dirs: Optional[ProjectDirs] = None,
+        ignore_patterns: Optional[List[str]] = None,
+    ):
+        super().__init__(
+            source,
+            part_src_dir,
+            cache_dir=cache_dir,
+            source_tag=source_tag,
+            source_branch=source_branch,
+            source_commit=source_commit,
+            source_checksum=source_checksum,
+            source_submodules=source_submodules,
+            source_depth=source_depth,
+            project_dirs=project_dirs,
+            ignore_patterns=ignore_patterns,
+        )
+
+        self._validate()
+
+    def _validate(self) -> None:
+        """Validate this source.
+
+        :raises: InvalidSourceOptions if any bad options are used.
+        """
+        bad_options = []
+        for tag in self._invalid_tags:
+            if getattr(self, tag.replace("-", "_")):
+                bad_options.append(tag)
+        if bad_options:
+            raise errors.InvalidSourceOptions(
+                source_type=self._source_type, options=bad_options
+            )
+
+    @override
+    def provision(
+        self,
+        dst: Path,
+        keep: bool = False,
+        src: Optional[Path] = None,
+    ) -> None:
+        """Extract rpm file contents to the part source dir."""
+        rpm_path = src or self.part_src_dir / os.path.basename(self.source)
+        command = ["rpm2archive", "-"]
```

Comment:
Any special reason to choose rpm2archive instead of rpm2cpio? The latter is probably faster/lighter given the data format used in the rpm file.

Reply:
Good question, and one I debated internally given precisely the fact that current RPMs are typically CPIO-based. 

In the end I decided on it for future-proofing and compatibility reasons. While the default format with rpm v4 is still a CPIO file, there are cases ([example](https://kojipkgs.fedoraproject.org//packages/python-graph-tool/2.29/2.fc32/aarch64/python3-graph-tool-2.29-2.fc32.aarch64.rpm)) where rpm2cpio fails and suggests rpm2archive intsead. Additionally, the rpm v6 spec looks as though it's not going to be compatible with rpm2cpio.

My initial version included the `--nocompression` parameter for rpm2archive, but alas that was only added in rpm 4.17 and focal only has 4.14. I considered adding detection logic (`rpm2cpio`, fall back to `rpm2archive --nocompression`, finally fall back to `rpm2archive`), but decided the extra complexity probably wasn't worth the performance gain.

Reply:
It would be interesting to add a comment about that otherwise the next person to look at that code may want to change it to rpm2cpio.

Reply:
Good point. Adding!